### PR TITLE
placement: check if rule can match any store when adding

### LINKED ISSUE #3299: placement: need more illegal label constraint check
## Bug Report

Follow #3293, separate the check part from that issue. These are checks that needs to be done when adding rules.

Known illegal cases so far:

1. label key not existed
2. `["+dc=dc1", "+dc=dc2"]`, contradicted label
3. `["+dc=dc1", "-dc=dc1"]`, contradicted label
4. No-match constraint
5. `["+dc=dc1", "+dc=dc1"]`, not necessarily an error, but it is duplicated

All these labels will pass the check successfully.

### PR DESCRIPTION
**Pull request by xhebox**
### What problem does this PR solve?

close #3299 

PD did not check if rules can match any TiKV instances on adding. For example, there may be one TiKV with `dc=bj` and another one with `dc=sh`. Then one rule of `+dc=nj` won't match any instance. It is invalid and useless to add it.

This PR adds a check in the function `adjustRule`, which means every rule will be validated before adding.

**NOTE**: `checkRule` function comes from the [commit](https://github.com/xhebox/pd/commit/a1488e9466a8a63ed8b57c6cdc36e8d86b824524).

Placement rules in SQL is not a released feature, so it is OK to break the compatibility. 

`store` information is used exclusively for changing/adding rules.

### Check List

Tests

- Unit test
- Integration test

### Release note

- Disallow adding placement rule that won't match any TiKV store

### CODE REVIEW

File: server/schedule/checker/rule_checker_test.go [hunk 1]
```diff
@@ -374,3 +384,116 @@ func (s *testRuleCheckerSuite) TestIssue3293(c *C) {
...
 	c.Assert(op.Desc(), Equals, "add-rule-peer")
 }
+
+func (s *testRuleCheckerSuite) TestIssue3299(c *C) {
+	s.cluster.AddLabelsStore(1, 1, map[string]string{"host": "host1"})
+	s.cluster.AddLabelsStore(2, 1, map[string]string{"dc": "sh"})
+	s.cluster.AddLeaderRegionWithRange(1, "", "", 1, 2)
+	s.ruleManager.SetStores(s.cluster.GetStores())
+
+	err := s.ruleManager.SetRule(&placement.Rule{
```
Comment (rleungx) [inline server/schedule/checker/rule_checker_test.go:394]: How about making the test cases shorter since they have some common values in Rule?

Comment (xhebox) [reply server/schedule/checker/rule_checker_test.go]: Done, that is the best I can do.

**Review by nolouch — COMMENTED** [review]
please update the comment and title. is it only update stores when change the rules?

**Comment by xhebox** [comment]
> please update the comment and title.

I don't what to update...

> is it only update stores when changing the rules?

Yes, I use `store` infomation exclusively for changing/adding rules.

File: server/schedule/placement/rule_manager.go [hunk 2]
```diff
@@ -659,3 +664,11 @@ func (m *RuleManager) SetKeyType(h string) *RuleManager {
 	m.keyType = h
 	return m
 }
+
+// SetStores will update stores info for adjustRule()
+func (m *RuleManager) SetStores(h []*core.StoreInfo) *RuleManager {
```
Comment (nolouch) [inline server/schedule/placement/rule_manager.go:669]: Its easy to forget `SetStores` before `SetRules`. I suggest passing in a closure to get stores when creating rule management. like: 
```
c.ruleManager = placement.NewRuleManager(c.storage, c.GetStores)
```

Comment (xhebox) [reply server/schedule/placement/rule_manager.go:669]: Nice!

**Review by nolouch — APPROVED** [review]
lgtm

**Review by HunDunDM — COMMENTED** [review]
What is this file `pkg/typeutil/sednSN3dL` used for?

File: server/schedule/placement/rule_manager.go [hunk 3]
```diff
@@ -41,13 +41,15 @@ type RuleManager struct {
...
-	keyType string
+	keyType   string
+	getStores func() []*core.StoreInfo
 }
 
 // NewRuleManager creates a RuleManager instance.
-func NewRuleManager(store *core.Storage) *RuleManager {
+func NewRuleManager(store *core.Storage, getStores func() []*core.StoreInfo) *RuleManager {
 	return &RuleManager{
 		store:      store,
```
Comment (HunDunDM) [inline server/schedule/placement/rule_manager.go:51]: Maybe we need to rename this to `storage` to avoid ambiguity?

**Review by HunDunDM — COMMENTED** [review]
rest LGTM

File: server/schedule/placement/rule_manager.go (hunk 3 above)
Comment (xhebox) [reply server/schedule/placement/rule_manager.go]: Done.

File: server/schedule/placement/rule_manager.go [hunk 4]
```diff
@@ -34,20 +34,22 @@ import (
...
 
 	// used for rule validation
-	keyType string
+	keyType   string
+	getStores func() []*core.StoreInfo
 }
 
 // NewRuleManager creates a RuleManager instance.
-func NewRuleManager(store *core.Storage) *RuleManager {
+func NewRuleManager(storage *core.Storage, getStores func() []*core.StoreInfo) *RuleManager {
```
Comment (Yisaer) [inline server/schedule/placement/rule_manager.go:49]: I think it's better to use `StoreSetInformer` interface here.

**Review by Yisaer — CHANGES_REQUESTED** [review]
Rest LGTM. plz remove `pkg/typeutil/sednSN3dL`

File: server/schedule/placement/rule_manager.go (hunk 4 above)
Comment (xhebox) [reply server/schedule/placement/rule_manager.go:49]: Addressed.

**Review by HunDunDM — APPROVED** [review]

**Review by nolouch — APPROVED** [review]

**Comment by nolouch** [comment]
cc @Yisaer

**Review by Yisaer — APPROVED** [review]
LGTM

**Pull request merged**